components/steering_wheel/src/IO/IO_Digital.c

```diff
@@ -0,0 +1,73 @@
+/**
+ * IO_Digital.c
+ * Digital IO Module source code
+ * @note To transfer this module to another system, the initialized inputs
+ *      of IO_DIGITAL must be modified to be congruent to the new system
+ */
+
+/******************************************************************************
+ *                             I N C L U D E S
+ ******************************************************************************/
+
+// Module Header
+#include "IO/IO_Digital.h"
+
+// System includes
+#include <string.h>
+
+// Other includes
+#include "ModuleDesc.h"
+#include "Utility.h"
+#include "SystemConfig.h"
+
+
+/******************************************************************************
+ *                           P U B L I C  V A R S
+ ******************************************************************************/
+
+IO_Digital_S IO_DIGITAL = {
+    .inputs[SW1] = { .gpiox = SW1_GPIO_Port, .pin = SW1_Pin },
+    .inputs[SW2] = { .gpiox = SW2_GPIO_Port, .pin = SW2_Pin },
+    .inputs[SW4] = { .gpiox = SW4_GPIO_Port, .pin = SW4_Pin },
+    .inputs[SW5] = { .gpiox = SW5_GPIO_Port, .pin = SW5_Pin },
+    .inputs[BTN_TOP_LEFT] = { .gpiox = BTN_TOP_LEFT_GPIO_Port, .pin = BTN_TOP_LEFT_Pin },
+    .inputs[BTN_MIDDLE] = { .gpiox = BTN_MIDDLE_GPIO_Port, .pin = BTN_MIDDLE_Pin },
+    .inputs[BTN_TOP_RIGHT] = { .gpiox = BTN_TOP_RIGHT_GPIO_Port, .pin = BTN_TOP_RIGHT_Pin }
+}; /**< Must Initialize amount of IO_Digital_Input_S equal to NUM_INPUTS */
+
+
+/******************************************************************************
+ *                       P U B L I C  F U N C T I O N S
+ ******************************************************************************/
+
+/**
+ * @brief Initializes the Digital IO Init
+ */
+static void IO_Digital_init(void)
+{
+    memset(&IO_DIGITAL.inputState, 0x00, sizeof(IO_Digital_S) - sizeof(IO_Digital_Input_S)*NUM_INPUTS);
+}
+
+/**
+ * @brief Periodic 10Hz function checking input states
+ */
+static void IO_Digital_10Hz_PRD(void)
+{
+    /**< Because of pull-up, signal is active low */
+    for (digitalInput_E i = (digitalInput_E) 0U; i < NUM_INPUTS; i++)
+    {
+        if (HAL_GPIO_ReadPin(IO_DIGITAL.inputs[i].gpiox, IO_DIGITAL.inputs[i].pin) == GPIO_PIN_RESET)
```

Comment:
perfect example of somewhere where it makes more sense to implement the function ourselves rather than use the HAL function so that we can customize it to our needs in the future. Here's what this function does in the HAL:
```
GPIO_PinState HAL_GPIO_ReadPin(GPIO_TypeDef *GPIOx, uint16_t GPIO_Pin)
{
  GPIO_PinState bitstatus;

  /* Check the parameters */
  assert_param(IS_GPIO_PIN(GPIO_Pin));

  if ((GPIOx->IDR & GPIO_Pin) != (uint32_t)GPIO_PIN_RESET)
  {
    bitstatus = GPIO_PIN_SET;
  }
  else
  {
    bitstatus = GPIO_PIN_RESET;
  }
  return bitstatus;
}
```

Reply:
Should it be standard throughout the whole project to interact with the hardware through registers (not considering initialisation type activities)? All my historical usage has been through HAL more or less so would be new to me but is faisable. When you say customize to our needs, could you give me a use case so I can better understand?

Reply:
see my other comment for example of what could be useful. Yeah the idea would be that we have our own API for accessing GPIO stuff, and we can implement it in different ways for each chip, but that will come down to the chip initialization stuff. Like what I did for the CAN layer, we just call CanSend for example, and the low level stuff gets done directly in can.c